### PR TITLE
feature: Introduce configurability to PhpdocOrderFixer

### LINKED ISSUE #1602: PhpdocOrderFixer - order?
Current PhpdocOrderFixer produce the following order:

``` php
/**
 * @param
 *
 * @throws
 *
 * @return
 */
```

What is the reason of keeping that order, not the param-return-throws ? Is is standardized by PSR?

### PR DESCRIPTION
**Pull request by kusab85**
Closes #1602.

Introduce configurability to PhpdocOrderFixer with option ``order`` which explicitly defines sequence of annotations.

Option ``order`` accepts array of strings.
Default value is  ``['param', 'throws', 'return']``.

### CODE REVIEW

File: src/Fixer/Phpdoc/PhpdocOrderFixer.php [hunk 1]
```diff
@@ -81,38 +143,43 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
...
+     *
+     * @param string   $move   Tag of annotations that should be moved
+     * @param string[] $before Tags of annotations that should moved annotations be placed before
      */
-    private function moveParamAnnotations(string $content): string
+    private function moveAnnotationsBefore(string $move, array $before, string $content): string
     {
         $doc = new DocBlock($content);
-        $params = $doc->getAnnotationsOfType('param');
+        $to_be_moved = $doc->getAnnotationsOfType($move);
```
Comment (SpacePossum) [inline src/Fixer/Phpdoc/PhpdocOrderFixer.php:168]: please make this camel cased :)

**Comment by SpacePossum** [comment]
Haven't tested it yet, but this PR looks very promising :) thanks for making it!

I wonder, would be more flexible if we would allow people to configure like `'order' =>  ['@throws', '@return', '@param']` where the annotations can be any order they want instead of picking one of the two sets?

File: src/Fixer/Phpdoc/PhpdocOrderFixer.php (hunk 1 above)
Comment (kusab85) [reply src/Fixer/Phpdoc/PhpdocOrderFixer.php]: done!

**Comment by kusab85** [comment]
> I wonder, would be more flexible if we would allow people to configure like `'order' => ['@throws', '@return', '@param']` where the annotations can be any order they want instead of picking one of the two sets?

I considered this option, but chose to stick with the predefined sets.
Maybe it's a bit stiff, but it makes people stick to the coding standards ;-)

Of course, there is also a third option. Implement two configuration options: 'style' and 'order'. Where 'style' allows you to pick an order according to a specific coding standard and 'order' allows you to configure any order you want.

Or maybe better way is to implement one configuration option 'order' that accepts string and array.
String lets people pick a specific coding standard  and array lets them configure order they want.

File: doc/list.rst [hunk 2]
```diff
@@ -2016,7 +2016,22 @@ List of Available Rules
    `Source PhpCsFixer\\Fixer\\Phpdoc\\PhpdocNoUselessInheritdocFixer <./../src/Fixer/Phpdoc/PhpdocNoUselessInheritdocFixer.php>`_
 -  `phpdoc_order <./rules/phpdoc/phpdoc_order.rst>`_
 
-   Annotations in PHPDoc should be ordered so that ``@param`` annotations come first, then ``@throws`` annotations, then ``@return`` annotations.
+   Annotations in PHPDoc should be ordered in specific style.
+
+   Annotations in PHPDoc should be ordered in one of the styles below:
+
+   - ``'phpcs'`` style annotations order is ``@param``, ``@throws``,
```
Comment (keradus) [inline doc/list.rst:2023]: CodeSniffer?

**Comment by keradus** [comment]
- reading `phpcs` some will think "Code Sniffer" ?
- reading `symfony` some will think "i never touch that framework. probably I don't want to check what is it"

can we have the names more self-explanatory ?

**Comment by kusab85** [comment]
@keradus are you suggesting getting rid of this no explanatory names and leaving just one option that accepts array with explicitly defined annotations order?

**Comment by keradus** [comment]
I'm OK with both, having an explicit option as an array, or having more descriptive option names

**Comment by jrmajor** [comment]
Another feature idea (I'm willing to prepare a separate PR after this one gets accepted): allow wildcards for cases like `@template*` to match `@template` and `@template-covariant`, `@psalm-*` to have all Psalm-specific docblocks grouped together, or `@*return` to have `@return`, `@psalm-return` and `@phpstan-return` grouped together.

**Comment by jrmajor** [comment]
Maybe something like `throws_before_return` and `throws_after_return` (or just a boolean option)? I would really encourage you to accept a list of ordered annotation names though :)

**Comment by kusab85** [comment]
Ok then, I'll go with 'throw_(before|after)_return' descriptive names and an optional explicit order in a form of ordered list of annotations names.

**Comment by kusab85** [comment]
There you go, people. Now we're left with one option ``'order'`` which lets you configure **any** order of **any** annotations you want. 

To support wildcards it requires modification of  PhpCsFixer\DocBlock\DocBlock class, so I think it'd be better to do it in separate PR.

And there's one more thing. I wonder if tests for the fixer should be more extensive. Please, let me know what you think. ;-)

File: doc/rules/phpdoc/phpdoc_order.rst [hunk 3]
```diff
@@ -28,10 +63,55 @@ Example #1
...
+     */
+
 Rule sets
 ---------
 
 The rule is part of the following rule set:
 
 @PhpCsFixer
-  Using the `@PhpCsFixer <./../../ruleSets/PhpCsFixer.rst>`_ rule set will enable the ``phpdoc_order`` rule.
+  Using the `@PhpCsFixer <./../../ruleSets/PhpCsFixer.rst>`_ rule set will enable the ``phpdoc_order`` rule with the default config.
```
Comment (VincentLanglet) [inline doc/rules/phpdoc/phpdoc_order.rst:117]: It should be added to the Symfony ruleset with the config:

```
@param, @return, @throws
```
according to
https://symfony.com/doc/current/contributing/code/standards.html

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
I think this PR will also close https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1602

Also, if custom annotation are supported, it will fix half of the issue:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5026

File: doc/rules/phpdoc/phpdoc_order.rst (hunk 3 above)
Comment (kusab85) [reply doc/rules/phpdoc/phpdoc_order.rst:117]: Suggestion applied!

Comment (kusab85) [reply doc/rules/phpdoc/phpdoc_order.rst:117]: Hey @VincentLanglet, I've got a test issue with your suggestion.

`RuleSetTest::testThatDefaultConfigIsNotPassed` throws an error:
`Rule "phpdoc_order" (in RuleSet "@ Symfony") has default config passed.`

It's all because the test **sorts** configuration arrays before comparing them.

Do you have any idea how to approach this problem?

For the record:
Default config for the fixer is: `['order' => ['param', 'throws', 'return']]`
@ Symfony's config for the fixer is: `['order' => ['param', 'return', 'throws']]`

Comment (VincentLanglet) [reply doc/rules/phpdoc/phpdoc_order.rst:117]: Since order has impact, this shouldn't be sorted.

Dunno if
- The sort should be removed
- An option should be added to not always sort for this test

but in your case this should definitively not be sorted.

Comment (kusab85) [reply doc/rules/phpdoc/phpdoc_order.rst:117]: Ok, I did it this way:

```
[...]
        // do not sort options for which order is important
        $excludeFromSorting = $this->orderMattersFor($ruleName);

        static::assertNotSame(
            $this->sortNestedArray($defaultConfig, $excludeFromSorting),
            $this->sortNestedArray($ruleConfig, $excludeFromSorting),
            sprintf('Rule "%s" (in RuleSet "%s") has default config passed.', $ruleName, $setName)
        );

[...]
```

```
[...]
    private function orderMattersFor(string $ruleName): array
    {
        static $orderMattersFor = [
            'phpdoc_order' => ['order'],
        ];

        return $orderMattersFor[$ruleName] ?? [];
    }
[...]
```

```
[...]
    private function sortNestedArray(array $array, array $excludeKeys = []): array
    {
        foreach ($array as $key => $element) {
            if (!\is_array($element) || \in_array($key, $excludeKeys, true)) {
                continue;
            }

            $array[$key] = $this->sortNestedArray($element, $excludeKeys);
        }
[...]
```

Comment (kusab85) [reply doc/rules/phpdoc/phpdoc_order.rst:117]: ... but I think it's wrong :)

Comment (kusab85) [reply doc/rules/phpdoc/phpdoc_order.rst:117]: > ``RuleSetTest::testThatDefaultConfigIsNotPassed`` throws an error:
> ``Rule "phpdoc_order" (in RuleSet "@Symfony") has default config passed.``
> 
> It's all because the test sorts configuration arrays before comparing them.

@SpacePossum or @jrmajor? Could you please shed some light on this matter? 
I'm stuck with it. I want to minimize code changes, but I am losing clarity of meaning.

For example, I can use the normalizer to convert ``array`` to ``ArrayIterator`` so that ``RuleSetTest`` will skip sorting the value of the ``order`` option. 
 
https://github.com/gdl2studio/PHP-CS-Fixer/blob/0650d49124ad7dc2347047cec9a90000d4b848dc/src/Fixer/Phpdoc/PhpdocOrderFixer.php#L111

File: tests/RuleSet/RuleSetTest.php [hunk 4]
```diff
@@ -419,23 +419,78 @@ public function testInvalidConfig(): void
...
+        if (\in_array($path, $orderMatters, true)) { // order matters
+            return;
         }
 
-        // sort by key if associative, by values otherwise
-        if (array_keys($array) === range(0, \count($array) - 1)) {
-            sort($array);
+        $keys = array_keys($data);
+
+        if ($this->allInteger($keys)) {
```
Comment (julienfalque) [inline tests/RuleSet/RuleSetTest.php:449]: Why this change?

**Review by julienfalque — COMMENTED** [review]
Looks good!

File: src/Fixer/Phpdoc/PhpdocOrderFixer.php [hunk 5]
```diff
@@ -54,12 +58,40 @@ public function getDefinition(): FixerDefinitionInterface
...
+    /**
+     * {@inheritdoc}
+     *
+     * @param array<string, mixed> $configuration
+     */
+    public function configure(array $configuration): void
+    {
+        parent::configure($configuration);
+
+        $this->order = (array) $this->configuration['order'];
```
[the remark this answers is not in the document]
Comment (kusab85) [reply src/Fixer/Phpdoc/PhpdocOrderFixer.php:84]: You're right, the ``$order`` is unnecessary. I removed it completely.

File: src/Fixer/Phpdoc/PhpdocOrderFixer.php [hunk 6]
```diff
@@ -81,38 +126,53 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
...
             $content = $token->getContent();
-            // move param to start, return to end, leave throws in the middle
-            $content = $this->moveParamAnnotations($content);
-            // we're parsing the content again to make sure the internal
-            // state of the docblock is correct after the modifications
-            $content = $this->moveReturnAnnotations($content);
+
+            // sort annotations
+            $successors = $this->order;
+            if (\count($successors) >= 2) {
```
[the remark this answers is not in the document]
Comment (kusab85) [reply src/Fixer/Phpdoc/PhpdocOrderFixer.php:134]: Checking this condition in this place makes no sense. 
I assured minimum ``$successors`` array length by adding value validity check in FixerOption
```php
// [...]
->setAllowedValues([function ($order) {
	if (\count($order) < 2) {
		throw new InvalidOptionsException('The option "order" value is invalid. Minimum two tags are required.');
	}
//[...]
```

File: tests/Fixer/Phpdoc/PhpdocOrderFixerTest.php [hunk 7]
```diff
@@ -208,4 +236,400 @@ public function testExampleFromSymfony(): void
...
+                'bool   $bar Bar',
+                'class-string<T> $id',
+            ],
+            '@template' => 'T of Extension\Extension',
+        ];
+
+        return [
+            [
+                ['order' => ['internal', 'template', 'param', 'custom', 'throws', 'return']],
+                self::glueBricks(
```
[the remark this answers is not in the document]
Comment (kusab85) [reply tests/Fixer/Phpdoc/PhpdocOrderFixerTest.php:542]: Good point. Test simplified!

File: tests/RuleSet/RuleSetTest.php [hunk 8]
```diff
@@ -419,23 +419,78 @@ public function testInvalidConfig(): void
...
+    /**
+     * Sorts an array of fixer definition recursively.
+     *
+     * Sometimes keys are all string, sometimes they are integers - we need to account for that.
+     *
+     * @param array<int|string,mixed> $data
+     */
+    private function doSort(array &$data, string $path): void
+    {
+        static $orderMatters = [
```
[the remark this answers is not in the document]
Comment (kusab85) [reply tests/RuleSet/RuleSetTest.php:438]: Moved to class private constant.

File: tests/RuleSet/RuleSetTest.php (hunk 4 above)
Comment (kusab85) [reply tests/RuleSet/RuleSetTest.php:449]: You mean this skipping of sorting an array of only arrays?
Ah, this change has strayed from a different branch. 
It shouldn't be here. At least yet.

File: src/Fixer/Phpdoc/PhpdocOrderFixer.php [hunk 9]
```diff
@@ -71,6 +87,26 @@ public function getPriority(): int
...
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('order', 'Sequence in which annotations in PHPDoc should be ordered.'))
+                ->setAllowedTypes(['string[]'])
+                ->setAllowedValues([function ($order) {
+                    if (\count($order) < 2) {
+                        throw new InvalidOptionsException('The option "order" value is invalid. Minimum two tags are required.');
```
Comment (julienfalque) [inline src/Fixer/Phpdoc/PhpdocOrderFixer.php:100]: Please add a test for this scenario.

Comment (kusab85) [reply src/Fixer/Phpdoc/PhpdocOrderFixer.php:100]: Test added.

**Review by julienfalque — APPROVED** [review]

**Comment by julienfalque** [comment]
Thank you @kusab85.

**Comment by kusab85** [comment]
🎉 Hooray! 🎉  
It's a huge satisfaction to make contribution to such a great tool. 
Thanks for merging @julienfalque !

**Comment by keradus** [comment]
really kudos for this PR, @kusab85 !

File: src/RuleSet/Sets/SymfonySet.php [hunk 10]
```diff
@@ -140,6 +140,13 @@ public function getRules(): array
             'phpdoc_no_alias_tag' => true,
             'phpdoc_no_package' => true,
             'phpdoc_no_useless_inheritdoc' => true,
+            'phpdoc_order' => [
+                'order' => [
```
Comment (keradus) [inline src/RuleSet/Sets/SymfonySet.php:144]: I hope this change was aligned with Sf community?

Comment (kusab85) [reply src/RuleSet/Sets/SymfonySet.php:144]: Not exactly ;-) How can I reach them?

Comment (julienfalque) [reply src/RuleSet/Sets/SymfonySet.php:144]: ping @nicolas-grekas or @derrabus Is this change OK or should we revert it?

**Comment by SpacePossum** [comment]
@kusab85 , very nice :D  thanks so much!

File: src/RuleSet/Sets/SymfonySet.php (hunk 10 above)
Comment (derrabus) [reply src/RuleSet/Sets/SymfonySet.php:144]: LGTM
